### ISSUE TITLE
[Storage] Uploading file using Storage doesn't work 

### ISSUE THREAD

**Issue by concretevitamin** [skypilot-org/skypilot#1226]
`~/tmpfile` exists as a file (tried both empty and non-empty).

```
resources:
  cloud: aws

file_mounts:
  ~/tmpfile:
    name: zongheng-tmp-1011
    source: ~/tmpfile

run: echo hi
```

```
» sky launch tmp.yaml
Task from YAML spec: tmp.yaml
I 10-11 12:36:19 storage.py:390] Detected existing storage object, loading Storage: zongheng-tmp-1011
I 10-11 12:36:20 storage.py:414] Verifying bucket and uploading from source for storage zongheng-tmp-1011
⠋ Syncing ~/tmpfile to s3://zongheng-tmp-1011/E 10-11 12:36:21 storage.py:857] warning: Skipping file /Users/zongheng/tmpfile/. File does not exist.
E 10-11 12:36:21 storage.py:857]
E 10-11 12:36:21 storage.py:672] Could not upload ~/tmpfile to store name zongheng-tmp-1011.
sky.exceptions.StorageUploadError: Upload to S3 failed for store zongheng-tmp-1011 and source ~/tmpfile. Please check the logs.
```

**Comment by romilbhardwaj** [comment]
ahh this is a duplicate of #1069:

TL;DR - 
1. If the destination is intended to be a file, we cannot support MOUNT mode because all mounting tools require the mount point to be an empty directory. In this case, goofys/s3fuse would need `~/` to be empty. 
    * While it might be possible to modify their source to use `mount -o nonempty`, it's tricky because the existing contents of the directory [would not be visible](https://askubuntu.com/questions/11269/are-pre-existing-files-in-a-mountpoint-directory-safe-and-uneffected) till the storage is unmounted.
2. We can allow only COPY mode, but we will need special handling. We will need to maintain and pass the source file name and the bucket name to the VM during file mounting to ensure only the specific file is copied to the VM (e.g., when the user uploads additional files to the bucket).

We can implement 2. Should discuss if that would solve the usecase.

If we choose not to implement 2., then we should at the minimum raise an error here, stating single file paths are not supported in Sky Storage.

**Comment by romilbhardwaj** [comment]
Some notes from offline discussion:

* To preface - the above example is definitely undesirable behavior and we should at the very least raise an error here, telling the user to either use direct file mounts OR to move their files into a directory.
* Sky Storage is intended to be a "box" where users can put files. As a result, destination for Sky Storage must always be a directory, not a file. 
* That said, one of our users' request is uploading multiple files to the same bucket and have them appear at a common directory.
* Should we add support for specifying a list of files for Storage? Something like:
  ```
  file_mounts:
    /mymount:
      name: romilb-bucket
      source: [/file1.txt, /file2.txt..]    # List of files to be uploaded to the bucket
      
  # Files show up at /mymount/file1.txt, /mymount/file2.txt and so on
  ```

**Comment by romilbhardwaj** [comment]
This is partially mitigated with #1231, but I feel we need to a longer term solution to supporting uploading files to Sky Storage (perhaps the list of files suggestion above?). Leaving the issue open.

**Comment by romilbhardwaj** [comment]
Duplicated in #1070, and likely will be closed with #1311.
